### PR TITLE
feat: diffusion backend

### PR DESCRIPTION
**Pull request by maharajamihir**

### CODE REVIEW

File: jasmine/utils/nn.py [hunk 1]
```diff
@@ -486,6 +482,357 @@ def __call__(
         return x_BTNV
 
 
+def modulate(x, shift, scale):
+    return x * (1 + scale) + shift
+
+
+# TODO mihir clean this up
```
[the remark this answers is not in the document]
Comment (emergenz) [reply jasmine/utils/nn.py:489]: @maharajamihir

File: jasmine/genie.py [hunk 2]
```diff
@@ -552,6 +610,140 @@ def generation_step_fn(
...
+                _rng_noise_context, (B, seq_len, N, L)
+            )
+            currupted_latents_BSNL = (
+                latents_BSNL * t_corrupt_context
+                + (1 - t_corrupt_context) * noise_context_BSNL
+            )
+            frame_mask = jnp.arange(seq_len) < frame_i
+            frame_mask_1S11 = frame_mask.reshape(1, seq_len, 1, 1)
+            corrupted_tok_latents_BSNL = jnp.where(
+                frame_mask_1S11, currupted_latents_BSNL, latents_BSNL
```
[the remark this answers is not in the document]
Comment (emergenz) [reply jasmine/genie.py:712]: @maharajamihir

[the remark this answers is not in the document]
Comment (emergenz) [reply jasmine/genie.py:712]: @maharajamihir

**Review by emergenz — COMMENTED** [review]
Adapt the README accordingly

File: jasmine/train_tokenizer_vqvae.py [hunk 3]
```diff
@@ -345,36 +343,41 @@ def tokenizer_loss_fn(
...
+        q_loss = jnp.square(jax.lax.stop_gradient(outputs["emb"]) - outputs["z"]).mean()
+        commitment_loss = jnp.square(
+            outputs["emb"] - jax.lax.stop_gradient(outputs["z"])
+        ).mean()
+        loss = mse + q_loss + args.vq_beta * commitment_loss
+        _, index_counts = jnp.unique_counts(
+            jnp.ravel(outputs["indices"]), size=args.num_latents, fill_value=0
+        )
+        codebook_usage = (index_counts != 0).mean()
+        metrics["loss"] = loss
```
Comment (emergenz) [inline jasmine/train_tokenizer_vqvae.py:367]: Create the metrics dict in one go like in main.

File: jasmine/train_tokenizer_vqvae.py [hunk 4]
```diff
@@ -402,13 +405,15 @@ def val_step(
...
-    def calculate_validation_metrics(val_dataloader, tokenizer):
+    def calculate_validation_metrics(val_dataloader, tokenizer, rng):
         step = 0
         loss_per_step = []
         metrics_per_step = []
         batch = None
         recon = None
         for batch in val_dataloader:
+            rng, _rng_mask = jax.random.split(rng, 2)
+            batch["rng"] = _rng_mask
```
Comment (emergenz) [inline jasmine/train_tokenizer_vqvae.py:416]: We don't actually use batch["rng"] in the VQ-VAE-based tokenizer, right?

File: jasmine/models/tokenizer.py [hunk 5]
```diff
@@ -128,3 +128,135 @@ def decode(self, indices_BTN: jax.Array, video_hw: Tuple[int, int]) -> jax.Array
...
+                lambda rng, prob: jax.random.bernoulli(rng, prob, (T, N)),
+                in_axes=(0, 0),
+            )(jnp.asarray(_rngs_mask), mask_prob)
+            patch_BTNP = jnp.where(
+                mask_BTN[..., None], self.mask_patch.value, patch_BTNP
+            )
+
+        # --- Encode ---
+        z_BTNL = self.encoder(patch_BTNP)
+        z_BTNL = nnx.tanh(z_BTNL)
```
Comment (emergenz) [inline jasmine/models/tokenizer.py:253]: Please add a note that this is from the Dreamer 4 paper.

File: jasmine/models/tokenizer.py [hunk 6]
```diff
@@ -128,3 +128,135 @@ def decode(self, indices_BTN: jax.Array, video_hw: Tuple[int, int]) -> jax.Array
...
+    ) -> Dict[str, jax.Array]:
+        # --- Preprocess videos ---
+        B, T = videos.shape[:2]
+        patch_BTNP = patchify(videos, self.patch_size)
+        N = patch_BTNP.shape[2]
+
+        # --- Randomly mask patches ---
+        if training:
+            _rng_prob, *_rngs_mask = jax.random.split(rng, B + 1)
+            mask_prob = jax.random.uniform(
```
Comment (emergenz) [inline jasmine/models/tokenizer.py:240]: This is a per-sequence masking probability. We should use per-frame masking probabilities if we want to match Dreamer 4.

**Comment by emergenz** [comment]
Closed in favour of https://github.com/p-doom/jasmine/pull/239
